jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -241,7 +239,7 @@ private void writeTarFile(Path outputPath) {
             () ->
                 new WriteTarFileStep(
                         buildConfiguration,
-                        childProgressDispatcherFactorySupplier.get(),
+                        childProgressDispatcherFactories.pop(),
```

Comment:
what am I missing here? Why can't this call just be: progressEventDispatcher.newChildProducer() ?

Reply:
It sounds like `newChildProducer()` shouldn't be called in separate threads, so we need to create them all in the main thread first and then pass them to each step when they run.

Reply:
Either way, I find this mechanism a little strange. Why not just keep a reference to progressEventDispatcher instead of pre-generating children?

Reply:
For this scheduling, if we just do `progressEventDispatcher.newChildProducer()`, it will not be called immediately because it is wrapped in a runnable as `() -> ... progressEventDispatcher.newChildProducer() ...`. Here, the main thread should create child producers and construct the tree properly.

Reply:
You were right @loosebazooka. I can directly call `progressEventDispatcher.newChildProducer()`; just need to do it at the beginning.